<a href="https://colab.research.google.com/github/Hannah1123/UTS_ML2019_ID13135684/blob/master/A2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Introduction  
定义问题， 意义

#Exploration  
challenges，设计数据结构，计划数据模型和测试  
Identify Challenges;
Design Data Structures/Plan Data Models and Tests
W-300

#Methodology  
Implement Algorithms;  
Build and Train Data Models  
W-300 for Practical Project

In [0]:
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import rcParams

In [0]:
link = 'https://drive.google.com/open?id=1CzKH7BCeewOyjsw0bNI7Gn76I-meIkzX'

In [7]:
fluff, id = link.split('=')
print (id) # Verify that i have everything after '='

1CzKH7BCeewOyjsw0bNI7Gn76I-meIkzX


In [0]:
downloaded = drive.CreateFile({'id':id}) 
downloaded.GetContentFile('Filename.csv')  
df3 = pd.read_csv('Filename.csv')
# Dataset is now stored in a Pandas Dataframe

In [9]:
df3.head()

url  ... listed_in(city)
0  https://www.zomato.com/bangalore/jalsa-banasha...  ...    Banashankari
1  https://www.zomato.com/bangalore/spice-elephan...  ...    Banashankari
2  https://www.zomato.com/SanchurroBangalore?cont...  ...    Banashankari
3  https://www.zomato.com/bangalore/addhuri-udupi...  ...    Banashankari
4  https://www.zomato.com/bangalore/grand-village...  ...    Banashankari

[5 rows x 17 columns]

In [0]:
del df3['url']
del df3['address']
del df3['phone']
del df3['location']

In [14]:
df3.head()

name online_order  ... listed_in(type) listed_in(city)
0                  Jalsa          Yes  ...          Buffet    Banashankari
1         Spice Elephant          Yes  ...          Buffet    Banashankari
2        San Churro Cafe          Yes  ...          Buffet    Banashankari
3  Addhuri Udupi Bhojana           No  ...          Buffet    Banashankari
4          Grand Village           No  ...          Buffet    Banashankari

[5 rows x 13 columns]

in the column 'rate', '/5' is not necessary.

In [0]:
df3.rate = df3.rate.replace("NEW", np.nan)
df3.dropna(how ='any', inplace = True)

In [16]:

df3.rate = df3.rate.astype(str)
df3.rate =df3.rate.apply(lambda x: x.replace('/5',''))
df3.rate = df3.rate.apply(lambda x: float(x))
df3.rename(columns={'approx_cost(for two people)': 'average_cost', 'listed_in(city)': 'locality'}, inplace=True)
df3.head()

name online_order  ... listed_in(type)      locality
0                  Jalsa          Yes  ...          Buffet  Banashankari
1         Spice Elephant          Yes  ...          Buffet  Banashankari
2        San Churro Cafe          Yes  ...          Buffet  Banashankari
3  Addhuri Udupi Bhojana           No  ...          Buffet  Banashankari
4          Grand Village           No  ...          Buffet  Banashankari

[5 rows x 13 columns]

In [17]:
len(df3)

23259

In [18]:
df3['rest_type'].unique()

array(['Casual Dining', 'Cafe, Casual Dining', 'Quick Bites',
       'Casual Dining, Cafe', 'Cafe', 'Quick Bites, Cafe', 'Delivery',
       'Dessert Parlor', 'Pub', 'Beverage Shop', 'Bar',
       'Takeaway, Delivery', 'Food Truck', 'Quick Bites, Dessert Parlor',
       'Pub, Casual Dining', 'Casual Dining, Bar', 'Bakery', 'Sweet Shop',
       'Dessert Parlor, Beverage Shop', 'Beverage Shop, Quick Bites',
       'Microbrewery, Casual Dining', 'Sweet Shop, Quick Bites', 'Lounge',
       'Food Court', 'Cafe, Bakery', 'Microbrewery', 'Kiosk', 'Pub, Bar',
       'Casual Dining, Pub', 'Cafe, Quick Bites', 'Lounge, Bar',
       'Bakery, Quick Bites', 'Dessert Parlor, Quick Bites',
       'Bar, Casual Dining', 'Beverage Shop, Dessert Parlor',
       'Casual Dining, Microbrewery', 'Mess', 'Lounge, Casual Dining',
       'Cafe, Dessert Parlor', 'Dessert Parlor, Cafe',
       'Bakery, Dessert Parlor', 'Quick Bites, Sweet Shop', 'Takeaway',
       'Microbrewery, Pub', 'Club', 'Fine Dining', 'Baker

In [0]:
df3['rest_type'].value_counts()

In [20]:
df3['listed_in(type)'].value_counts()

Delivery              10664
Dine-out               8009
Desserts               1364
Cafes                  1066
Drinks & nightlife      879
Buffet                  754
Pubs and bars           523
Name: listed_in(type), dtype: int64

In [0]:
cafe_data = df3.loc[df3['listed_in(type)'] == 'Cafes']

In [22]:
cafe_data.head(5)

name  ...      locality
7                                              Onesta  ...  Banashankari
8                                      Penthouse Cafe  ...  Banashankari
9                                           Smacznego  ...  Banashankari
10  Café Down The A...  ...  Banashankari
11                                       Cafe Shuffle  ...  Banashankari

[5 rows x 13 columns]

the best cafes   
cheapest  
highly rated  
reliable(large number of votes)options

In [23]:
cafe_data.average_cost.describe()
maxi=cafe_data.average_cost.max()
print('max of average cost in cafes:',maxi)
mean=cafe_data.rate.mean()
print('mean rate of cafes:', mean)

max of average cost in cafes: 950
mean rate of cafes: 4.027016885553465


cheapest: the cafes with average cost 1/4th the average cost of most expensive cafe in the dataframe.  
1/4th of 950 is 237.5

In [0]:
cafe_data['test']= cafe_data['average_cost'].str.replace(',','').astype(float)

In [0]:
cafe_data= cafe_data.drop_duplicates(subset='name',keep='first')
dups_name = cafe_data.pivot_table(index=['name'],aggfunc='size')
newcafe_data=cafe_data[['name','test','locality','rest_type','cuisines']].groupby(['test'], sort = True)



In [0]:
#filter1 = filter(lambda x: x <= 237,newcafe_data['average_cost'])
#newcafe_data['test']= newcafe_data['average_cost'].str.replace(',','').astype(float)
#newcafe_data=newcafe_data.filter(lambda x: x.mean() <= 237,newcafe_data)
#print(filter1)
newcafe_data=newcafe_data.filter(lambda x: x['test'].mean() <= 237)



In [66]:
#newcafe_data=pd.DataFrame(list(filter1))

#newcafe_data=filter1.sort_values(by=['average_cost'])
newcafe_data=newcafe_data.sort_values(by=['test'])
newcafe_data

"""

targets1 = cafe_data[(cafe_data['rate'] >= 4.5) & (cafe_data['average_cost'].mean() <= 237) & (cafe_data['votes'] >= 175)].sort_values('average_cost')
targets2 = cafe_data[(cafe_data['rate'] >= 4.5) & (cafe_data['average_cost'].mean() >= 450) & (cafe_data['votes'] >= 175)].sort_values('average_cost')
"""


"\n\ntargets1 = cafe_data[(cafe_data['rate'] >= 4.5) & (cafe_data['average_cost'].mean() <= 237) & (cafe_data['votes'] >= 175)].sort_values('average_cost')\ntargets2 = cafe_data[(cafe_data['rate'] >= 4.5) & (cafe_data['average_cost'].mean() >= 450) & (cafe_data['votes'] >= 175)].sort_values('average_cost')\n"

In [0]:
newcafe_data_expensive=cafe_data[['name','test','locality','rest_type','cuisines']].groupby(['test'], sort = True)
newcafe_data_expensive=newcafe_data_expensive.filter(lambda x: x['test'].mean() >= 560)
newcafe_data_expensive=newcafe_data_expensive.sort_values(by=['test'])

In [69]:
newcafe_data_rate=cafe_data[['name','rate']].groupby(['rate'], sort = True)
newcafe_data_rate=newcafe_data_rate.filter(lambda x: x['rate'].mean() >= 4.5)
newcafe_data_rate=newcafe_data_rate.sort_values(by=['rate'])
cafe_data.rate.value_counts()
cafe_data.rate.unique()
cafe_data.nunique()

name               338
online_order         2
book_table           2
rate                23
votes              261
rest_type           19
dish_liked         332
cuisines           207
average_cost        26
reviews_list       336
menu_item           83
listed_in(type)      1
locality            25
test                26
dtype: int64

In [72]:
newcafe_data

name   test  ... rest_type                                  cuisines
2527   4th Tea Block  150.0  ...      Cafe                                      Cafe
8307     Kullad Cafe  200.0  ...      Cafe  North Indian, Cafe, Fast Food, Beverages
8328   Fabindia Cafe  200.0  ...      Cafe                         Cafe, Street Food
17573      Chai Days  200.0  ...      Cafe           Cafe, Tea, Beverages, Fast Food
17609      Chai Cafe  200.0  ...      Cafe                                      Cafe

[5 rows x 5 columns]

#Evaluation  
Report Execution on Data;  
Perform and Report Testing;  
Perform Efficiency Analysis;  
Do Possible Comparative Study  
W-200

#Conclusion  
Discuss Reflections;  
Propose Possible Improvements  
W-100

#Ethical